frontend/src/metabase-types/api/dataset.ts

```diff
@@ -144,6 +144,7 @@ export interface TemplateTag {
   dimension?: LocalFieldReference;
   "widget-type"?: string;
   required?: boolean;
+  optional: boolean;
```

Comment:
I honestly don't see how this thing isn't the same as `required`, but negated.

Reply:
The `required` prop is not set from parsing the query, but from the parameter that is linked to the template tag.

I feel this is confusing and we should work on separating the two so that parameters never write back into template tags.

The reason I created a new property (`optional`) is that I wanted to add this context without having to change how parameters work currently.
Adding a property allows us to migrate away from the current way things work 
incrementally: `optional` informs us about what the state of the query text is,
`required` informs us what the state of the linked parameter is (and we can 
potentially remove this in the future or merge both properties).

The [MLv2 type for a template tag](https://github.com/metabase/metabase/blob/master/src/metabase/lib/schema/template_tag.cljc#L113-L122) does not actually have a `:required` property, so I picked one that didn't clash with current `TemplateTag` properties.

Reply:
After the in-person discussion, I see that `optional` should mean "is within the double square brackets `[[ ]]`" while `required` is about the requires switch state for this template tag. This is ok.

However, I am not sure the parser architecture should be this as well as the way we decide whether or not the query is runnable as-is should depend on the `optional` flag.

Let's discuss with the product first, as agreed.

Reply:
Can we please add JSDocs explaining the difference?